File: test6_printLinklistTailtoHead.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
"""
File: test6_printLinklistTailtoHead.py
Author: heguilong
Email: hgleagle@gmail.com
Github: https://github.com/hgleagle
Description:
输入一个链表，从尾到头打印链表每个节点的值。
"""
from __future__ import print_function


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def __init__(self):
        self.reverse_list = []

    def printListFromTailToHead(self, listNode):
        """ 从尾部到头部的列表值序列，例如[1,2,3]

        递归或者栈实现

        Returns: None

        """
        if listNode is None:
            return
        else:
            self.printListFromTailToHead(listNode.next)
        self.reverse_list.append(listNode.val)


if __name__ == "__main__":
    prev = None
    head = None
    for i in range(10):
        node = ListNode(i)
        if head is None:
            head = node
        if prev:
            prev.next = node
        prev = node
    sol = Solution()
    sol.printListFromTailToHead(head)
    print(sol.reverse_list)

File: test6_printLinklistTailtoHead.c
/*
 * 输入一个链表，从尾到头打印链表每个节点的值。
 * 用栈实现更好,不建议下面这个实现
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


typedef struct ListNode {
      int val;
      struct ListNode *next;
} ListNode_t;


void printListFromTailToHead(ListNode_t *head)
{
    ListNode_t *next_node = head;
    int length = 0;
    int i = 0;

    while (next_node != NULL) {
        next_node = next_node->next;
        length += 1;
    }

    int value[length];
    memset(value, 0, length);
    next_node = head;
    while (i < length) {
        value[i] = next_node->val;
        next_node = next_node->next;
        i++;
    }

    for (i = length - 1; i >= 0; i--) {
        printf("%d\n", value[i]);
    }
}

int main(void)
{
    ListNode_t *head, *node, *prev;
    int i;

    prev = NULL;
    head = NULL;
    for (i = 0; i < 10; i++) {
        node = (ListNode_t *)malloc(sizeof(ListNode_t));
        if (node == NULL) {
            break;
        }
        node->val = i;
        node->next = NULL;
        if (prev != NULL) {
            prev->next = node;
        }
        if (head == NULL) {
            head = node;
        }
        prev = node;
    }

    if (head != NULL) {
        printListFromTailToHead(head);
    }

    node = head;
    while (node) {
        prev = node;
        free(node);
        node = prev->next;
    }
}
